File: Python/binarytree.py
from node import Node
from collections import deque
class BinaryTree:
    def __init__(self):
        self.root = None
    def insert(self, val):
        new_node = Node(val)
        if self.root == None:
            self.root = new_node
        else:
            cur = self.root
            while True:
                if new_node.val > cur.val:
                    if not cur.right:
                        cur.right = new_node
                        break
                    else:
                        cur = cur.right
                elif new_node.val < cur.val:
                    if not cur.left:
                        cur.left = new_node
                        break
                    else:
                        cur = cur.left
                else:
                    raise KeyError("Cannot insert same value")
    def contains(self, val):
        if not self.root:
            return False
        else:
            cur = self.root
            while cur:
                if cur.val == val:
                    return True
                elif val < cur.val:
                    cur = cur.left
                else:
                    cur = cur.right
            return False
    def BFS(self):
        que = deque([self.root])
        visited = []
        while que:
            level = len(que)
            for i in range(level):
                cur = que.popleft()
                visited.append(cur.val)
                if cur.left : que.append(cur.left)
                if cur.right : que.append(cur.right)
        return ",".join([str(x) for x in visited])
    def preorder(self):
        arr = []
        def traverse(node):
            arr.append(node.val)
            if node.left : traverse(node.left)
            if node.right : traverse(node.right)
        traverse(self.root)
        return " , ".join([str(x) for x in arr])
    def inorder(self):
        arr = []
        def traverse(node):
            if node.left : traverse(node.left)
            arr.append(node.val)
            if node.right : traverse(node.right)
        traverse(self.root)
        return " , ".join([str(x) for x in arr])
    def postorder(self):
        arr = []
        def traverse(node):
            if node.left : traverse(node.left)
            if node.right : traverse(node.right)
            arr.append(node.val)
        traverse(self.root)
        return " , ".join([str(x) for x in arr])

tr = BinaryTree()
tr.insert(98)
tr.insert(78)
tr.insert(-98)
tr.insert(12)
tr.insert(76)
print(tr.inorder())

File: C++/binarytree.cpp
#include <iostream>
#include "binarytree.h"
#include <queue>
#include <vector>
#include <functional>
using std::queue;
using std::vector;
using std::endl;
using std::function;
using std::cout;
binarytree::binarytree()
{
    root = nullptr;
}

binarytree::~binarytree()
{
    if(!root) {return;}
    queue<node*> que;
    node* temp = root;
    que.push(temp);
    while(!que.empty()) {
        node* cur = que.front();
        que.pop();
        if(cur->left != nullptr) {
            que.push(cur->left);
        }
        if(cur->right != nullptr) {
            que.push(cur->right);
        }
        delete cur;
    }
}
void binarytree::push(int value) {
    node* new_node = new node(value);
    if(root == nullptr) {
        root = new_node;
    } else {
        node* cur = root;
        while(true) {
            if(cur->value < new_node->value) {
                if(cur->right == nullptr) {
                    cur->right = new_node;
                    break;
                } else {
                    cur = cur->right;
                }
            } else if(cur->value > new_node->value) {
                if(cur->left == nullptr) {
                    cur->left = new_node;
                    break;
                } else {
                    cur = cur->left;
                }
            } else {
                std::cout << "UNACCEOPTABLE" << std::endl;
                break;
            }
        }
    }
}
bool binarytree::contains(int val) {
    if(root == nullptr) return false;
    node* temp = root;
    while(temp != nullptr) {
        if(temp->value == val) {
            return true;
        } else if(temp->value < val) {
            temp = temp->right;
        } else {
            temp = temp->left;
        }
    }
    return false;
}
vector<int> binarytree::preorder() {
    vector<int> arr = {};
    function<void (node*)> traverse;
    traverse = [&](node* temp) {
        if(temp == nullptr) {
            return;
        }
        arr.push_back(temp->value);
        traverse(temp->left);
        traverse(temp->right);
    };
    traverse(root);
    return arr;
}
vector<int> binarytree::inorder() {
    vector<int> arr;
    function<void (node*)> traverse = [&](node* temp) {
        if(temp == nullptr) return;
        traverse(temp->left);
        arr.push_back(temp->value);
        traverse(temp->right);
    };
    traverse(root);
    return arr;
}
vector<int> binarytree::postorder() {
    vector<int> arr;
    function<void(node*)> traverse = [&](node* node) {
        if(node == nullptr) {
            return;
        }
        traverse(node->left);
        traverse(node->right);
        arr.push_back(node->value);
    };
    traverse(root);
    return arr;
}
vector<int> binarytree::BFS() {
    vector<int> visited;
    queue<node*> que;
    que.push(root);
    while(!que.empty()) {
        node* temp = que.front();
        que.pop();
        visited.push_back(temp->value);
        if(temp->left) {que.push(temp->left);}
        if(temp->right) {que.push(temp->right);}
    }
    return visited;
}